apps/OboeTester/app/src/main/java/com/mobileer/oboetester/ReverseJniActivity.java

```diff
@@ -0,0 +1,137 @@
+
+import static com.mobileer.oboetester.TestAudioActivity.TAG;
+
+import android.content.Context;
+import android.media.AudioManager;
+import android.os.Bundle;
+import android.os.Handler;
+import android.os.Looper;
+import android.util.Log;
+import android.view.View;
+import android.view.WindowManager;
+import android.widget.Button;
+import android.widget.CheckBox;
+import android.widget.RadioButton;
+import android.widget.Spinner;
+import android.widget.TextView;
+
+import androidx.appcompat.app.AppCompatActivity;
+
+/**
+ * Test for reverse jni
+ */
+public class ReverseJniActivity extends AppCompatActivity {
+
+    private ReverseJniEngine mAudioEngine;
+    private Button mStartButton;
+    private Button mStopButton;
+    TextView mStatusTextView;
+    private ExponentialSliderView mSleepDurationUsSlider;
+    private ExponentialSliderView mBufferSizeInBurstsSlider;
+    private boolean mIsPlaying = false;
+
+    private Handler mUiUpdateHandler;
+    private Runnable mUiUpdateRunnable;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_reverse_jni); // Assumes a basic layout with two buttons
+
+        mAudioEngine = new ReverseJniEngine();
+        mStartButton = findViewById(R.id.startButton);
+        mStopButton = findViewById(R.id.stopButton);
+        mStatusTextView = findViewById(R.id.text_status);
+        mSleepDurationUsSlider = findViewById(R.id.sleep_duration_us);
+        mBufferSizeInBurstsSlider = findViewById(R.id.num_bursts);
+
+        mStartButton.setOnClickListener(v -> {
+            if (!mIsPlaying) {
+                startAudio();
+            }
+        });
+
+        mStopButton.setOnClickListener(v -> {
+            if (mIsPlaying) {
+                stopAudio();
+            }
+        });
+
+        mSleepDurationUsSlider.setOnValueChangedListener(new ExponentialSliderView.OnValueChangedListener() {
+            @Override
+            public void onValueChanged(ExponentialSliderView view, int value) {
+                mAudioEngine.setSleepDurationUs(value);
+            }
+        });
+
+        mBufferSizeInBurstsSlider.setOnValueChangedListener(new ExponentialSliderView.OnValueChangedListener() {
+            @Override
+            public void onValueChanged(ExponentialSliderView view, int value) {
+                mAudioEngine.setBufferSizeInBursts(value);
+            }
+        });
+
+        mUiUpdateHandler = new Handler(Looper.getMainLooper());
+        mUiUpdateRunnable = new Runnable() {
+            @Override
+            public void run() {
+                if (mIsPlaying && mAudioEngine != null) {
+                    int xruns = mAudioEngine.getXRunCount();
+                    // This must be run on the UI thread
+                    mStatusTextView.setText(String.format("XRuns: %d", xruns));
+                    mUiUpdateHandler.postDelayed(this, 50); // Update every 50ms
```

Comment:
Given the data callback will also update the UI, why it is needed to update every 50ms?